### PR TITLE
PDOSWorkChain - align energy range to fermi level

### PR DESCRIPTION
**Pull request by t-reents**
Hi,
I implemented the changes that were suggested in this comment https://github.com/aiidateam/aiida-quantumespresso/issues/716#issuecomment-909201940.

I already talked about this PR with @mbercx. I am not familiar with pytest, so I did not include additional tests. I would be happy to get some help concerning the tests.

### CODE REVIEW

File: aiida_quantumespresso/workflows/pdos.py [hunk 1]
```diff
@@ -219,14 +222,15 @@ def define(cls, spec):
...
         )
         spec.input(
-            'align_to_fermi',
-            valid_type=orm.Bool,
+            'fermi_energy_range',
+            valid_type=orm.List,
+            required=False,
             serializer=to_aiida_type,
-            default=lambda: orm.Bool(False),
+            default=lambda: orm.List(list=[]),
```
Comment (sphuber) [inline aiida_quantumespresso/workflows/pdos.py:229]: Don't think this needs a default if it is going to be empty anyway.

File: aiida_quantumespresso/workflows/pdos.py [hunk 2]
```diff
@@ -219,14 +222,15 @@ def define(cls, spec):
             help='Terminate workchain steps before submitting calculations (test purposes only).'
         )
         spec.input(
-            'align_to_fermi',
-            valid_type=orm.Bool,
+            'fermi_energy_range',
+            valid_type=orm.List,
```
Comment (sphuber) [inline aiida_quantumespresso/workflows/pdos.py:226]: This also needs a validator to check that when specified, it contains two float values

File: aiida_quantumespresso/workflows/pdos.py [hunk 3]
```diff
@@ -219,14 +222,15 @@ def define(cls, spec):
             help='Terminate workchain steps before submitting calculations (test purposes only).'
         )
         spec.input(
-            'align_to_fermi',
-            valid_type=orm.Bool,
+            'fermi_energy_range',
```
Comment (sphuber) [inline aiida_quantumespresso/workflows/pdos.py:225]: I might be saying something stupid because I wasn't part of the discussions, but does the name `fermi_energy_range` really make sense? The values define the energy window in which the DOS is computed. Does something like just `energy_range` or `energy_window` not make more sense?

File: aiida_quantumespresso/workflows/pdos.py [hunk 4]
```diff
@@ -452,10 +456,16 @@ def _generate_dos_inputs(self):
...
-            dos_parameters['DOS']['Emin'] = dos_parameters['Emin'] + self.ctx.nscf_fermi
-            dos_parameters['DOS']['Emax'] = dos_parameters['Emax'] + self.ctx.nscf_fermi
+        if fermi_energy_range:
+            dos_parameters['DOS']['Emin'] = fermi_energy_range[0] + self.ctx.nscf_fermi
+            dos_parameters['DOS']['Emax'] = fermi_energy_range[1] + self.ctx.nscf_fermi
+        elif 'Emin' in dos_parameters['DOS'] and 'Emax' in dos_parameters['DOS']:
+            pass
+        else:
+            dos_parameters['DOS']['Emin'] = self.ctx.nscf_emin
+            dos_parameters['DOS']['Emax'] = self.ctx.nscf_emax
```
Comment (sphuber) [inline aiida_quantumespresso/workflows/pdos.py:468]:
```suggestion
        else:
            dos_parameters['DOS'].setdefault('Emin', self.ctx.nscf_emin)
            dos_parameters['DOS'].setdefault('Emax', self.ctx.nscf_emax)
```
This solution might be a bit more elegant although it is not completely identical since now the keys are decoupled and won;t be updated only if neither are specified. But this may actually be a desirable side effect.

**Review by sphuber — COMMENTED** [review]
Thanks @t-reents . The changes seem ok in principle, just have some suggestions.

File: aiida_quantumespresso/workflows/pdos.py (hunk 1 above)
Comment (t-reents) [reply aiida_quantumespresso/workflows/pdos.py:229]: I agree, I will remove this

File: aiida_quantumespresso/workflows/pdos.py (hunk 4 above)
Comment (t-reents) [reply aiida_quantumespresso/workflows/pdos.py:468]: Thanks @sphuber. I was not aware of this. I will add this solution.

File: aiida_quantumespresso/workflows/pdos.py (hunk 3 above)
Comment (t-reents) [reply aiida_quantumespresso/workflows/pdos.py:225]: I see your point. If a energy window is specified by this parameter, the DOS/PDOS is computed in this range with respect to the fermi energy. For example, the range 0 - 30 eV is specified and the fermi energy is 7 eV, then the DOS would be computed in the range 7 - 37 eV. Therefore, I think it would be useful to indicate this also by the name of the input parameter (but maybe the description in the help keyword would be already enough).

File: aiida_quantumespresso/workflows/pdos.py (hunk 2 above)
Comment (t-reents) [reply aiida_quantumespresso/workflows/pdos.py:237]: I added a validator. I split it into two parts. Maybe one would prefer to combine it by an `or` and return just one messsage.

File: aiida_quantumespresso/workflows/pdos.py [hunk 5]
```diff
@@ -115,10 +115,13 @@ def validate_inputs(value, _):
...
             return f'The `{par}`` parameter has to be equal for the `dos` and `projwfc` inputs.'
 
-    if value.get('align_to_fermi', False):
+    if value.get('fermi_energy_range', False):
         for par in ['Emin', 'Emax']:
-            if value['dos']['parameters']['DOS'].get(par, None) is None:
-                return f'The `{par}`` parameter must be set in case `align_to_fermi` is set to `True`.'
+            if value['dos']['parameters']['DOS'].get(par, None):
+                warnings.warn(
+                    f'The `{par}`` parameter and fermi_energy_range were specified.'
```
Comment (sphuber) [inline aiida_quantumespresso/workflows/pdos.py:122]:
```suggestion
                    f'The `{par}` parameter and `fermi_energy_range` were specified.'
```

File: aiida_quantumespresso/workflows/pdos.py [hunk 6]
```diff
@@ -115,10 +115,13 @@ def validate_inputs(value, _):
...
+                    'The value in fermi_energy_range will be used.'
```
Comment (sphuber) [inline aiida_quantumespresso/workflows/pdos.py:123]:
```suggestion
                    'The value in `fermi_energy_range` will be used.'
```

File: aiida_quantumespresso/workflows/pdos.py [hunk 7]
```diff
@@ -159,6 +162,17 @@ def validate_projwfc(value, _):
...
 
+def validate_fermi_energy_range(value):
+    """Validate specified fermi_energy_range.
+    
+    - List needs to consist of two float values.
+    """
+    if len(value) != 2:
+        return 'List needs to consist of two values.'
+    if not all(isinstance(val, float) for val in value):
+        return 'Fermi energy range needs to be specified as floats.'
```
Comment (sphuber) [inline aiida_quantumespresso/workflows/pdos.py:173]:
```suggestion
    if len(value) != 2:
        return f'`fermi_energy_range` should be a `List` of length two, but got: {value}'
    if not all(isinstance(val, float) for val in value):
        return f'`fermi_energy_range` should be a `List` of floats, but got: {value}'
```

File: aiida_quantumespresso/workflows/pdos.py (hunk 3 above)
Comment (sphuber) [reply aiida_quantumespresso/workflows/pdos.py:236]: To me it reads a bit weird having it in the option, so I would vote to put it simply in the description that it is relative to the fermi level. I will let @mbercx have the final vote.

File: aiida_quantumespresso/workflows/pdos.py (hunk 2 above)
Comment (sphuber) [reply aiida_quantumespresso/workflows/pdos.py:237]: Validator looks good, just a suggestion on the error message to make it a bit clearer and more consistent

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @t-reents , just some final minor nitpicks that I noticed. Then I let @mbercx decide on the naming of the input. Once that is decided and applied we can merge this

**Comment by t-reents** [comment]
Thanks again @sphuber. I committed your suggestions.

File: aiida_quantumespresso/workflows/pdos.py [hunk 8]
```diff
@@ -159,6 +162,17 @@ def validate_projwfc(value, _):
     jsonschema.validate(value['parameters'].get_dict()['PROJWFC'], get_parameter_schema())
 
 
+def validate_fermi_energy_range(value, _):
```
Comment (t-reents) [inline aiida_quantumespresso/workflows/pdos.py:165]: @sphuber I added "_" in the inputs of the validator. Otherwise one would get a warning that a validator with just one input is deprecated. Please correct me if I am wrong.

Comment (sphuber) [reply aiida_quantumespresso/workflows/pdos.py:165]: That is the correct solution indeed, we do this routinely.

File: aiida_quantumespresso/workflows/pdos.py (hunk 3 above)
Comment (mbercx) [reply aiida_quantumespresso/workflows/pdos.py:236]: Sorry for being later to the party! Although I agree the name isn't perfect, I still think it's better to highlight that it's an energy range around the Fermi level by adding it to the name. Maybe `energy_range_fermi` would give less of an impression that there is some range of Fermi levels considered?

Comment (t-reents) [reply aiida_quantumespresso/workflows/pdos.py:236]: Personally, I like the last one but I leave this up to you

Comment (sphuber) [reply aiida_quantumespresso/workflows/pdos.py:236]: I also vote for the latter :+1:  A few more characters is almost always preferable if it adds clarity. So @t-reents please feel free to go with that one, update the PR and we can merge. Thanks!

Comment (t-reents) [reply aiida_quantumespresso/workflows/pdos.py]: The changes are now included

File: aiida_quantumespresso/workflows/pdos.py [hunk 9]
```diff
@@ -452,10 +467,14 @@ def _generate_dos_inputs(self):
         dos_inputs = AttributeDict(self.exposed_inputs(DosCalculation, 'dos'))
         dos_inputs.parent_folder = self.ctx.nscf_parent_folder
         dos_parameters = self.inputs.dos.parameters.get_dict()
+        fermi_energy_range = self.inputs.fermi_energy_range
```
Comment (sphuber) [inline aiida_quantumespresso/workflows/pdos.py:470]: I think you forgot to update it here

Comment (t-reents) [reply aiida_quantumespresso/workflows/pdos.py]: Oh yes, I am sorry. Now there is no `fermi_energy_range` left

**Review by sphuber — APPROVED** [review]
Thanks a lot @t-reents Good to go.

**Comment by t-reents** [comment]
Thanks a lot for your comments and instructions! @sphuber

**Comment by sphuber** [comment]
The tests are failing. Apparently the validator is being called even if no value is specified. We could solve this by explicitly checking for this value in the validator and in that case just returning, but really this should not be happening. I am looking into `plumpy` which is the library responsible for this part of the code on why it is calling the validator regardless. It is really weird because the logic does seem to prevent this from happening and there are unit tests for it. But apparently, the test here in `aiida-quantumespresso` has some edge condition. Will report back once I know more.

**Comment by mbercx** [comment]
Tests pass, but maybe we should add some more to properly test the functionality with and without the `energy_range_vs_fermi` input?

Also: maybe we should move the long module docstring into the documentation page of the work chain? Right now it's a bit empty:

https://aiida-quantumespresso--764.org.readthedocs.build/en/764/howto/workflows/pdos.html

**Comment by t-reents** [comment]
> Tests pass, but maybe we should add some more to properly test the functionality with and without the `energy_range_vs_fermi` input?
> 
> Also: maybe we should move the long module docstring into the documentation page of the work chain? Right now it's a bit empty:
> 
> https://aiida-quantumespresso--764.org.readthedocs.build/en/764/howto/workflows/pdos.html

For sure, I will add the testst, soon.

**Comment by mbercx** [comment]
@t-reents cheers! Could you also rebase the branch on `origin/main`?
=

**Comment by t-reents** [comment]
Done!

File: src/aiida_quantumespresso/workflows/pdos.py [hunk 10]
```diff
@@ -383,6 +397,9 @@ def setup(self):
         """Initialize context variables that are used during the logical flow of the workchain."""
         self.ctx.serial_clean = 'serial_clean' in self.inputs and self.inputs.serial_clean.value
         self.ctx.dry_run = 'dry_run' in self.inputs and self.inputs.dry_run.value
+        self.ctx.energy_range_vs_fermi = (
```
Comment (mbercx) [inline src/aiida_quantumespresso/workflows/pdos.py:400]: Why do we have to add this to the context? Is it updated somehow?

**Review by mbercx — COMMENTED** [review]
Thanks @t-reents! Just a few questions and comments, and then we can merge this after the final commit on `main` is fixed to have a better message.

File: tests/conftest.py [hunk 11]
```diff
@@ -879,8 +879,8 @@ def _generate_workchain_ph(exit_code=None, inputs=None, return_inputs=False):
 def generate_workchain_pdos(generate_workchain, generate_inputs_pw, fixture_code):
     """Generate an instance of a `PdosWorkChain`."""
 
-    def _generate_workchain_pdos():
-        from aiida.orm import Bool, Dict
+    def _generate_workchain_pdos(emin=None, emax=None, energy_range_vs_fermi=None):
```
[the remark this answers is not in the document]
Comment (t-reents) [reply tests/conftest.py:882]: Yeah, I agree. Indeed, I thought that this is the easiest solution for now. Let's open an issue and we can discuss afterwards

File: src/aiida_quantumespresso/workflows/pdos.py (hunk 10 above)
Comment (t-reents) [reply src/aiida_quantumespresso/workflows/pdos.py:400]: No, I think I just did at some point because it is used in the method to generate the inputs for the PDOS and in the one for the DOS inputs. But as it's not updated, I can also simply access the `self.inputs` directly. Will update this aspect and let you know so we can merge it

Comment (t-reents) [reply src/aiida_quantumespresso/workflows/pdos.py]: Hi @mbercx, I just pushed the changes, should be fine now

**Review by mbercx — APPROVED** [review]
LGTM! Thanks a lot @t-reents

File: tests/conftest.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mbercx) [reply tests/conftest.py:882]: https://github.com/aiidateam/aiida-quantumespresso/issues/1068
